File: test/stream.js
const express = require('express');
const NodeWebcam = require('node-webcam');
const app = express();

// 웹캠 설정
const webcamOptions = {
    width: 1280,
    height: 720,
    quality: 100,
    frames: 60,
    output: "jpeg",
    device: false,
    callbackReturn: "buffer"
};

const Webcam = NodeWebcam.create(webcamOptions);

// MJPEG 스트림 생성을 위한 함수
function generateFrames(res) {
    const boundary = 'frame';
    res.writeHead(200, {
        'Content-Type': `multipart/x-mixed-replace; boundary=${boundary}`,
        'Cache-Control': 'no-cache',
        'Connection': 'close',
        'Pragma': 'no-cache'
    });

    // 주기적으로 프레임 캡처 및 전송
    const captureFrame = () => {
        Webcam.capture('', (err, data) => {
            if (err) {
                console.error('웹캠 캡처 오류:', err);
                return;
            }
            res.write(`--${boundary}\r\n`);
            res.write('Content-Type: image/jpeg\r\n');
            res.write('Content-Length: ' + data.length + '\r\n');
            res.write('\r\n');
            res.write(data);
            res.write('\r\n');
        });
    };

    // 60fps로 프레임 전송
    const interval = setInterval(captureFrame, 1000 / 60);

    // 클라이언트 연결이 종료되면 인터벌 정리
    res.on('close', () => {
        clearInterval(interval);
    });
}

// 스트리밍 라우트
app.get('/stream', (req, res) => {
    generateFrames(res);
});

// 서버 시작
const PORT = 5000;
app.listen(PORT, '0.0.0.0', () => {
    console.log(`서버가 포트 ${PORT}에서 실행 중입니다.`);
}); 

File: python/stream.py
from flask import Flask, Response
import cv2

app = Flask(__name__)

# 웹캠 초기화
camera = cv2.VideoCapture(0)  # 노트북 기본 웹캠 (ID: 0)


def generate_frames():
    while True:
        success, frame = camera.read()
        if not success:
            break
        else:
            # 프레임을 JPEG로 인코딩
            _, buffer = cv2.imencode(".jpg", frame)
            frame = buffer.tobytes()
            # MJPEG 스트림 형식으로 반환
            yield (b"--frame\r\n" b"Content-Type: image/jpeg\r\n\r\n" + frame + b"\r\n")


@app.route("/stream")
def video_feed():
    # 클라이언트가 MJPEG 스트림을 받을 수 있도록 설정
    return Response(
        generate_frames(), mimetype="multipart/x-mixed-replace; boundary=frame"
    )


if __name__ == "__main__":
    app.run(host="0.0.0.0", port=8080, debug=True)
